File: go/pkg/2023/day1.go
package solver2023

import (
	"fmt"
	"regexp"
	"strconv"
)

type SolverOne struct {
}

func (solver SolverOne) SolutionOne(lines []string) int {
    answer := 0
    regExp, _ := regexp.Compile("[0-9]")

    for _, element := range lines {
        if element == "" {
            continue
        }

        numbers := regExp.FindAllString(element, -1)

        if len(numbers) == 0 {
            panic("No numbers found")
        }

        number := fmt.Sprintf("%s%s", numbers[0], numbers[len(numbers)-1])
        convertedNumber, err := strconv.Atoi(number)

        if err != nil {
            panic("Could not make number")
        }

        answer += convertedNumber
    }

    return answer;
}

func (solver SolverOne) SolutionTwo(lines []string) int {
    answer := 0
    regExp, _ := regexp.Compile("[0-9]|one|two|three|four|five|six|seven|eight|nine|zero")

    for _, element := range lines {
        if element == "" {
            continue
        }

        numbers := regExp.FindAllString(element, -1)

        if len(numbers) == 0 {
            panic("No numbers found")
        }

        firstNumber := solver.nameToNumber(numbers[0])
        lastNumber := solver.nameToNumber(numbers[len(numbers) - 1])

        number := fmt.Sprintf("%s%s", firstNumber, lastNumber)
        convertedNumber, err := strconv.Atoi(number)

        if err != nil {
            panic("Could not make number")
        }

        answer += convertedNumber
    }

    return answer
}

func (solver SolverOne) processLine(line string) int {
    regExp, _ := regexp.Compile("[0-9]")

    numbers := regExp.FindAllString(line, -1)

    if len(numbers) == 0 {
        panic("No numbers found")
    }

    firstNumber := numbers[0]
    lastNumber := numbers[len(numbers) - 1]

    number := fmt.Sprintf("%s%s", firstNumber, lastNumber)
    convertedNumber, err := strconv.Atoi(number)

    if err != nil {
        panic("Could not make number")
    }

    return convertedNumber
}

func (solver SolverOne) nameToNumber(number string) string {
    switch number {
    case "one":
    number = "1"
    case "two":
        number = "2"
    case "three":
    number = "3"
    case "four":
        number = "4"
    case "five":
        number = "5"
    case "six":
        number = "6"
    case "seven":
        number = "7"
    case "eight":
        number = "8"
    case "nine":
        number = "9"
    case "zero":
        number = "0"
    }

    return number
}

File: rust/aoc/src/solver/s23/day1.rs
use std::mem::replace;

use crate::solver::solver::{find_all_by_regex, Solver};


pub struct Day1{
}

impl Solver for Day1 {
    fn solution_one(&self, lines: Vec<String>) -> i64 {
        let mut answer: i64 = 0;

        for line in lines {
            let (left, right) = get_first_and_last_number(line);

            let final_string = format!("{left}{right}");
            answer += final_string.parse::<i64>().unwrap();
        }

        return answer;
    }

    fn solution_two(&self, lines: Vec<String>) -> i64 {
        let mut answer: i64 = 0;

        for line in lines {
            let modified_line = replace_illegal(line);
            let (left, right) = get_first_and_last_number(modified_line);
            let final_string = format!("{left}{right}");
            answer += final_string.parse::<i64>().unwrap();
        }

        return answer;
    }
}

fn get_first_and_last_number(line: String) -> (i64, i64) {
    let numbers = find_all_by_regex(line, r"[0-9]");

    let first_number = numbers[0].parse::<i64>().unwrap();
    let last_number = numbers[numbers.len()-1].parse::<i64>().unwrap();
    
    return (first_number, last_number);
}

fn replace_illegal(line: String) -> String {
    return line.replace("oneight", "18")
        .replace("nineight", "98")
        .replace("twone", "21")
        .replace("threeight", "38")
        .replace("sevenine", "79")
        .replace("eightwo", "82")
        .replace("eighthree", "83")
        .replace("one", "1")
        .replace("two", "2")
        .replace("three", "3")
        .replace("four", "4")
        .replace("five", "5")
        .replace("six", "6")
        .replace("seven", "7")
        .replace("eight", "8")
        .replace("nine", "9")
        .to_string();
}
